File: ada/ada2012/Python3/SymbolTable.py
from Symbol import Symbol
from TypeClassification import TypeClassification


class SymbolTable:
    def __init__(self):
        self._scope_stack = []
        self._block_counter = 0
        global_scope = Symbol()
        global_scope.name = "global"
        global_scope.classification = {TypeClassification.Global_}
        self._scope_stack.append(global_scope)

        # Predefined scalar types
        for name in ["integer", "natural", "positive", "float", "long_float",
                      "character", "wide_character", "wide_wide_character",
                      "boolean", "duration"]:
            self._define_type(name, False)

        # Predefined composite types
        for name in ["string", "wide_string", "wide_wide_string"]:
            self._define_type(name, True)

        # Predefined exceptions
        for name in ["constraint_error", "program_error", "storage_error",
                      "tasking_error", "numeric_error"]:
            sym = Symbol()
            sym.name = name
            sym.classification = {TypeClassification.ExceptionName_}
            sym.predefined = True
            self.define(sym)

        # Predefined enumeration literals
        for name in ["true", "false"]:
            sym = Symbol()
            sym.name = name
            sym.classification = {TypeClassification.EnumerationLiteral_}
            sym.predefined = True
            self.define(sym)

    def _define_type(self, name, is_composite):
        sym = Symbol()
        sym.name = name
        sym.classification = {TypeClassification.TypeName_}
        sym.predefined = True
        sym.is_composite = is_composite
        self.define(sym)

    def enter_scope(self, new_scope):
        if self._scope_stack and new_scope is self._scope_stack[-1]:
            return
        self._scope_stack.append(new_scope)

    def exit_scope(self):
        self._scope_stack.pop()
        if not self._scope_stack:
            raise RuntimeError("SymbolTable: scope stack underflow")

    def current_scope(self):
        if not self._scope_stack:
            return None
        return self._scope_stack[-1]

    def define(self, symbol):
        current = self.current_scope()
        return self.define_in_scope(current, symbol)

    def define_in_scope(self, current_scope, symbol):
        symbol.name = symbol.name.lower()
        if symbol.name in current_scope.members:
            return False
        symbol.parent = current_scope
        current_scope.members[symbol.name] = symbol
        return True

    def resolve(self, name, start_scope=None):
        name = name.lower()
        if start_scope is None:
            for scope in reversed(self._scope_stack):
                if name in scope.members:
                    return scope.members[name]
            return None
        else:
            return start_scope.members.get(name)

    def push_block_scope(self):
        self._block_counter += 1
        block_scope = Symbol()
        block_scope.name = "block" + str(self._block_counter)
        block_scope.classification = {TypeClassification.Block_}
        block_scope.predefined = True
        self.enter_scope(block_scope)
        return block_scope

    def pop_block_scope(self):
        self.exit_scope()

    def __str__(self):
        if self._scope_stack:
            return self._to_string_helper(self._scope_stack[0], 0)
        return ""

    def _to_string_helper(self, scope, depth):
        result = ""
        indent = "  " * depth
        for sym in scope.members.values():
            if not sym.predefined:
                result += indent + str(sym) + "\n"
            if (TypeClassification.Block_ in sym.classification or
                TypeClassification.SubprogramName_ in sym.classification or
                TypeClassification.PackageName_ in sym.classification):
                result += self._to_string_helper(sym, depth + 1)
        return result

File: ada/ada2012/CSharp/SymbolTable.cs
using System;
using System.Collections.Generic;
using System.Text;

public class SymbolTable {
    private Stack<Symbol> scopeStack = new Stack<Symbol>();
    private int blockCounter = 0;

    public SymbolTable() {
        var globalScope = new Symbol() { Name = "global", Classification = new HashSet<TypeClassification>() {TypeClassification.Global_} };
        scopeStack.Push(globalScope);

        // Predefined scalar types
        DefineType("integer", false);
        DefineType("natural", false);
        DefineType("positive", false);
        DefineType("float", false);
        DefineType("long_float", false);
        DefineType("character", false);
        DefineType("wide_character", false);
        DefineType("wide_wide_character", false);
        DefineType("boolean", false);
        DefineType("duration", false);

        // Predefined composite types
        DefineType("string", true);
        DefineType("wide_string", true);
        DefineType("wide_wide_string", true);

        // Predefined exceptions
        Define(new Symbol() { Name = "constraint_error", Classification = new HashSet<TypeClassification>(){TypeClassification.ExceptionName_}, Predefined = true });
        Define(new Symbol() { Name = "program_error", Classification = new HashSet<TypeClassification>(){TypeClassification.ExceptionName_}, Predefined = true });
        Define(new Symbol() { Name = "storage_error", Classification = new HashSet<TypeClassification>(){TypeClassification.ExceptionName_}, Predefined = true });
        Define(new Symbol() { Name = "tasking_error", Classification = new HashSet<TypeClassification>(){TypeClassification.ExceptionName_}, Predefined = true });
        Define(new Symbol() { Name = "numeric_error", Classification = new HashSet<TypeClassification>(){TypeClassification.ExceptionName_}, Predefined = true });

        // Predefined enumeration literals
        Define(new Symbol() { Name = "true", Classification = new HashSet<TypeClassification>(){TypeClassification.EnumerationLiteral_}, Predefined = true });
        Define(new Symbol() { Name = "false", Classification = new HashSet<TypeClassification>(){TypeClassification.EnumerationLiteral_}, Predefined = true });
    }

    private void DefineType(string name, bool isComposite) {
        Define(new Symbol() { Name = name, Classification = new HashSet<TypeClassification>(){TypeClassification.TypeName_}, Predefined = true, IsComposite = isComposite });
    }

    public void EnterScope(Symbol newScope) {
        var current = scopeStack.Peek();
        if (newScope == current) return;
        scopeStack.Push(newScope);
    }

    public void ExitScope() {
        scopeStack.Pop();
        if (scopeStack.Count == 0)
            throw new Exception("SymbolTable: scope stack underflow");
    }

    public Symbol CurrentScope()
    {
        if (scopeStack.Count == 0) return null;
        return scopeStack.Peek();
    }

    public bool Define(Symbol symbol) {
        var currentScope = CurrentScope();
        return this.DefineInScope(currentScope, symbol);
    }

    public bool DefineInScope(Symbol currentScope, Symbol symbol) {
        // Case-insensitive: normalize name to lowercase
        symbol.Name = symbol.Name.ToLowerInvariant();
        if (currentScope.Members.ContainsKey(symbol.Name)) {
            return false;
        }
        symbol.Parent = currentScope;
        currentScope.Members[symbol.Name] = symbol;
        return true;
    }

    public Symbol Resolve(string name, Symbol startScope = null)
    {
        // Case-insensitive: normalize to lowercase
        name = name.ToLowerInvariant();
        if (startScope == null)
        {
            foreach (Symbol scope in scopeStack)
            {
                if (scope.Members.TryGetValue(name, out Symbol symbol))
                {
                    return symbol;
                }
            }
            return null;
        }
        else
        {
            if (startScope.Members.TryGetValue(name, out Symbol symbol))
            {
                return symbol;
            }
            return null;
        }
    }

    public Symbol PushBlockScope()
    {
        var blockScope = new Symbol()
        {
            Name = "block" + (++blockCounter),
            Classification = new HashSet<TypeClassification>() { TypeClassification.Block_ },
            Predefined = true
        };
        EnterScope(blockScope);
        return blockScope;
    }

    public void PopBlockScope()
    {
        ExitScope();
    }

    public List<Symbol> GetExportedSymbols()
    {
        var result = new List<Symbol>();
        var global = scopeStack.ToArray()[scopeStack.Count - 1]; // bottom of stack = global scope
        foreach (var entry in global.Members)
        {
            if (!entry.Value.Predefined)
                result.Add(entry.Value);
        }
        return result;
    }

    public override string ToString()
    {
        StringBuilder sb = new StringBuilder();
        var scopes = scopeStack.ToArray();
        if (scopes.Length > 0)
        {
            ToStringHelper(sb, scopes[scopes.Length - 1], 0);
        }
        return sb.ToString();
    }

    private void ToStringHelper(StringBuilder sb, Symbol scope, int depth)
    {
        string indent = new string(' ', depth * 2);
        foreach (var entry in scope.Members)
        {
            var sym = entry.Value;
            if (!sym.Predefined)
            {
                sb.AppendLine(indent + sym.ToString());
            }
            if (sym.Classification.Contains(TypeClassification.Block_) ||
                sym.Classification.Contains(TypeClassification.SubprogramName_) ||
                sym.Classification.Contains(TypeClassification.PackageName_))
            {
                ToStringHelper(sb, sym, depth + 1);
            }
        }
    }
}
